File: executor/menubar-macos/Sources/Core/WorkspaceFolder.swift
import Foundation

/// Adding and removing the named folders an executor may read.
///
/// Like `PermittedProgram`, this is a *pre*-check so the reach surface can
/// refuse in words rather than in an exit status.
/// `assertExecutorWorkspaceFolders` in the CLI decides for real — the grammar,
/// the duplicate and nesting rules, and whether a path is an ordinary directory
/// at all — and its answer is the one that lands in the policy.
public enum WorkspaceFolder {
    /// `EXECUTOR_WORKSPACE_FOLDER_MAXIMUM` in packages/schemas/src/executor.ts.
    public static let maximumCount = 16
    /// `EXECUTOR_WORKSPACE_FOLDER_NAME_MAXIMUM_LENGTH`.
    public static let maximumNameLength = 40
    /// `EXECUTOR_RESERVED_WORKSPACE_FOLDER_NAMES`. Windows turns these basenames
    /// into devices wherever a path is opened, and a reviewed policy has to stay
    /// legal on the next machine that loads it.
    public static let reservedNames: Set<String> = [
        "aux", "con", "nul", "prn",
        "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8", "com9",
        "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
    ]
    public static let fallbackName = "workspace"

    public static let nameRefusal =
        "A workspace folder name is 1 to \(maximumNameLength) lowercase letters, digits and interior "
        + "hyphens. It starts every path an agent writes, so it carries no slash, no dot and no case."

    /// `EXECUTOR_WORKSPACE_FOLDER_NAME_PATTERN`: `^[a-z0-9](?:[a-z0-9-]{0,38}[a-z0-9])?$`.
    public static func nameIsLegal(_ value: String) -> Bool {
        guard !value.isEmpty, value.count <= maximumNameLength else { return false }
        guard !reservedNames.contains(value) else { return false }
        let allowed = CharacterSet(charactersIn: "abcdefghijklmnopqrstuvwxyz0123456789-")
        guard value.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { return false }
        guard let first = value.first, let last = value.last else { return false }
        return first != "-" && last != "-"
    }

    /// The name a chosen directory gets, derived the way
    /// `deriveExecutorWorkspaceFolderName` derives it, so the name a person is
    /// offered here is the name the CLI would pick. The derivation is total:
    /// anything the grammar refuses becomes `workspace`.
    public static func derivedName(for path: String) -> String {
        let basename = (path.replacingOccurrences(of: "\\", with: "/") as NSString).lastPathComponent
        var hyphenated = ""
        var pendingHyphen = false
        for character in basename.lowercased() {
            if character.isASCII && (character.isLetter || character.isNumber) {
                if pendingHyphen && !hyphenated.isEmpty { hyphenated.append("-") }
                pendingHyphen = false
                hyphenated.append(character)
            } else {
                pendingHyphen = true
            }
        }
        let derived = String(hyphenated.prefix(maximumNameLength))
        return nameIsLegal(derived) ? derived : fallbackName
    }

    public static func validate(
        adding path: String,
        named requested: String?,
        to existing: [ExecutorDescription.Folder]
    ) -> Result<[ExecutorDescription.Folder], ExecutorRefusal> {
        let trimmed = path.trimmingCharacters(in: .whitespacesAndNewlines)
        guard trimmed.hasPrefix("/") else {
            return .failure(ExecutorRefusal("A workspace folder is an absolute path on this Mac."))
        }
        let typed = (requested?.trimmingCharacters(in: .whitespacesAndNewlines)).flatMap {
            $0.isEmpty ? nil : $0
        }
        let name = typed ?? derivedName(for: trimmed)
        guard nameIsLegal(name) else {
            return .failure(ExecutorRefusal(nameRefusal))
        }
        guard existing.count + 1 <= maximumCount else {
            return .failure(ExecutorRefusal(
                "An executor reaches at most \(maximumCount) folders."
            ))
        }
        guard !existing.contains(where: { $0.name == name }) else {
            return .failure(ExecutorRefusal(
                "This executor already reaches a folder named \(name). Two folders with one name "
                    + "would make every path starting with it ambiguous."
            ))
        }
        // Nesting is refused by the CLI because a file inside two folders would
        // have two workspace paths, which makes a receipt ambiguous about what
        // was read. Saying so here saves a person the round trip.
        for folder in existing where contains(folder.path, trimmed) || contains(trimmed, folder.path) {
            return .failure(ExecutorRefusal(
                folder.path == trimmed
                    ? "This executor already reaches \(trimmed), as \(folder.name)."
                    : "\(trimmed) and \(folder.path) contain one another. A file inside both would "
                        + "have two workspace paths, so only one of them can be a folder."
            ))
        }
        return .success(
            (existing + [ExecutorDescription.Folder(name: name, path: trimmed)])
                .sorted { $0.name < $1.name }
        )
    }

    /// Removing the last folder is refused: an executor paired against nothing
    /// can read nothing, and the CLI refuses an empty list too.
    public static func validate(
        removing name: String,
        from existing: [ExecutorDescription.Folder]
    ) -> Result<[ExecutorDescription.Folder], ExecutorRefusal> {
        let remaining = existing.filter { $0.name != name }
        guard remaining.count != existing.count else {
            return .failure(ExecutorRefusal("\(name) is not one of this executor's folders."))
        }
        guard !remaining.isEmpty else {
            return .failure(ExecutorRefusal(
                "An executor reaches at least one folder. Add the folder it should read before "
                    + "removing this one."
            ))
        }
        return .success(remaining)
    }

    private static func contains(_ outer: String, _ inner: String) -> Bool {
        outer == inner || inner.hasPrefix(outer.hasSuffix("/") ? outer : outer + "/")
    }
}

public extension ExecutorDescription {
    /// What the reach surface says about guest sessions, in the person's terms.
    /// A guest VM mounts one workspace, so this is a refusal to state rather
    /// than a detail to hide — the alternative is `command.run` binding to
    /// whichever folder happened to come first.
    var guestSessionNote: String {
        switch reach.guestSessions {
        case .available:
            return "A sandboxed command or coding session can start: exactly one folder is configured."
        case .refusedMultipleFolders:
            return "Sandboxed commands and coding sessions refuse to start while more than one folder "
                + "is configured — a guest mounts one workspace, so it would otherwise bind to "
                + "whichever folder came first. Listing, reading, writing and reviewing files work "
                + "across all of them."
        }
    }
}

File: executor/tray-windows/src-tauri/src/workspace_folder.rs
//! Adding and removing the named folders an executor may read.
//!
//! Like the permitted-command grammar, this is a *pre*-check so the reach
//! surface can refuse in words rather than in an exit status.
//! `assertExecutorWorkspaceFolders` in the CLI decides for real — the grammar,
//! the duplicate and nesting rules, and whether a path is an ordinary directory
//! at all — and its answer is the one that lands in the policy.

use crate::description::Folder;

/// `EXECUTOR_WORKSPACE_FOLDER_MAXIMUM` in packages/schemas/src/executor.ts.
pub const MAXIMUM_COUNT: usize = 16;
/// `EXECUTOR_WORKSPACE_FOLDER_NAME_MAXIMUM_LENGTH`.
pub const MAXIMUM_NAME_LENGTH: usize = 40;
/// `EXECUTOR_RESERVED_WORKSPACE_FOLDER_NAMES`. Windows turns these basenames
/// into devices wherever a path is opened, and a reviewed policy has to stay
/// legal on the next machine that loads it.
pub const RESERVED_NAMES: &[&str] = &[
    "aux", "con", "nul", "prn", "com1", "com2", "com3", "com4", "com5", "com6", "com7", "com8",
    "com9", "lpt1", "lpt2", "lpt3", "lpt4", "lpt5", "lpt6", "lpt7", "lpt8", "lpt9",
];
pub const FALLBACK_NAME: &str = "workspace";

pub const NAME_REFUSAL: &str =
    "A workspace folder name is 1 to 40 lowercase letters, digits and interior \
     hyphens. It starts every path an agent writes, so it carries no slash, no dot and no case.";

/// `EXECUTOR_WORKSPACE_FOLDER_NAME_PATTERN`: `^[a-z0-9](?:[a-z0-9-]{0,38}[a-z0-9])?$`.
pub fn name_is_legal(value: &str) -> bool {
    if value.is_empty() || value.len() > MAXIMUM_NAME_LENGTH {
        return false;
    }
    if RESERVED_NAMES.contains(&value) {
        return false;
    }
    let mut chars = value.chars();
    let Some(first) = chars.next() else { return false };
    let Some(last) = value.chars().last() else { return false };
    if first == '-' || last == '-' {
        return false;
    }
    value
        .chars()
        .all(|character| character.is_ascii_lowercase() || character.is_ascii_digit() || character == '-')
}

/// The name a chosen directory gets, derived the way
/// `deriveExecutorWorkspaceFolderName` derives it, so the name a person is
/// offered here is the name the CLI would pick. The derivation is total:
/// anything the grammar refuses becomes `workspace`.
pub fn derived_name(path: &str) -> String {
    let normalised = path.replace('\\', "/");
    let basename = std::path::Path::new(&normalised)
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or(FALLBACK_NAME);
    let mut hyphenated = String::new();
    let mut pending_hyphen = false;
    for character in basename.to_lowercase().chars() {
        if character.is_ascii_alphanumeric() {
            if pending_hyphen && !hyphenated.is_empty() {
                hyphenated.push('-');
            }
            pending_hyphen = false;
            hyphenated.push(character);
        } else {
            pending_hyphen = true;
        }
    }
    let derived = hyphenated.chars().take(MAXIMUM_NAME_LENGTH).collect::<String>();
    if name_is_legal(&derived) { derived } else { FALLBACK_NAME.to_owned() }
}

/// Validates adding one folder to an existing list. The caller may supply a
/// name or ask for the derived one.
pub fn validate_adding(
    path: &str,
    requested_name: Option<&str>,
    existing: &[Folder],
) -> Result<Vec<Folder>, String> {
    let trimmed = path.trim();
    if !std::path::Path::new(trimmed).is_absolute() {
        return Err("A workspace folder is an absolute path on this computer.".to_owned());
    }
    let name = requested_name
        .map(|value| value.trim())
        .filter(|value| !value.is_empty())
        .map(|value| value.to_owned())
        .unwrap_or_else(|| derived_name(trimmed));
    if !name_is_legal(&name) {
        return Err(NAME_REFUSAL.to_owned());
    }
    if existing.len() + 1 > MAXIMUM_COUNT {
        return Err(format!("An executor reaches at most {MAXIMUM_COUNT} folders."));
    }
    if existing.iter().any(|folder| folder.name == name) {
        return Err(format!(
            "This executor already reaches a folder named {name}. Two folders with one name \
             would make every path starting with it ambiguous."
        ));
    }
    for folder in existing {
        if contains(&folder.path, trimmed) || contains(trimmed, &folder.path) {
            return Err(if folder.path == trimmed {
                format!("This executor already reaches {trimmed}, as {}.", folder.name)
            } else {
                format!(
                    "{trimmed} and {} contain one another. A file inside both would have two \
                     workspace paths, so only one of them can be a folder.",
                    folder.path
                )
            });
        }
    }
    let mut folders = existing.to_owned();
    folders.push(Folder { name, path: trimmed.to_owned() });
    folders.sort_by(|left, right| left.name.cmp(&right.name));
    Ok(folders)
}

/// Removing the last folder is refused: an executor paired against nothing can
/// read nothing, and the CLI refuses an empty list too.
pub fn validate_removing(name: &str, existing: &[Folder]) -> Result<Vec<Folder>, String> {
    let remaining: Vec<Folder> = existing.iter().filter(|folder| folder.name != name).cloned().collect();
    if remaining.len() == existing.len() {
        return Err(format!("{name} is not one of this executor's folders."));
    }
    if remaining.is_empty() {
        return Err(
            "An executor reaches at least one folder. Add the folder it should read before \
             removing this one."
                .to_owned(),
        );
    }
    Ok(remaining)
}

fn contains(outer: &str, inner: &str) -> bool {
    let outer_with_separator = if outer.ends_with('/') || outer.ends_with('\\') {
        outer.to_owned()
    } else {
        format!("{outer}/")
    };
    outer == inner || inner.starts_with(&outer_with_separator)
}

#[cfg(test)]
mod tests {
    use super::{derived_name, name_is_legal, validate_adding, validate_removing, Folder};

    fn folder(name: &str, path: &str) -> Folder {
        Folder { name: name.to_owned(), path: path.to_owned() }
    }

    #[test]
    fn legal_names_match_the_schema_pattern() {
        assert!(name_is_legal("nessie"));
        assert!(name_is_legal("nessie-workspace"));
        assert!(name_is_legal("a1"));
        assert!(!name_is_legal(""));
        assert!(!name_is_legal("-nessie"));
        assert!(!name_is_legal("nessie-"));
        assert!(!name_is_legal("nessie.workspace"));
        assert!(!name_is_legal("Nessie"));
        assert!(!name_is_legal("con"));
        assert!(!name_is_legal(&"a".repeat(41)));
    }

    #[test]
    fn derived_name_turns_a_basename_into_a_legal_name() {
        assert_eq!(derived_name("/home/person/Nessie Workspace"), "nessie-workspace");
        assert_eq!(derived_name("/home/person/work"), "work");
        assert_eq!(derived_name("/weird/weird--name"), "weird-name");
    }

    #[test]
    fn adding_requires_an_absolute_path() {
        assert!(validate_adding("work", None, &[]).is_err());
        assert!(validate_adding("./work", None, &[]).is_err());
        assert!(validate_adding("/home/person/work", None, &[]).is_ok());
        // On Windows the same check accepts a drive-letter path; on this macOS
        // host `std::path::is_absolute` does not, so the assertion is gated.
        #[cfg(windows)]
        assert!(validate_adding("C:\\Users\\person\\work", None, &[]).is_ok());
    }

    #[test]
    fn duplicate_names_and_nesting_are_refused() {
        let existing = vec![folder("nessie", "/home/person/nessie")];
        assert!(validate_adding("/home/person/nessie", None, &existing).is_err());
        assert!(validate_adding("/home/person/nessie/src", None, &existing).is_err());
    }

    #[test]
    fn removing_the_last_folder_is_refused() {
        let existing = vec![folder("nessie", "/home/person/nessie")];
        assert!(validate_removing("nessie", &existing).is_err());
        assert!(validate_removing("missing", &existing).is_err());
    }
}
